### PR TITLE
Magic editor patch dark theme

### PR DESCRIPTION
**Pull request by rschrieken**

### CODE REVIEW

**Comment by dippas** [comment]
Related #185

File: Magic™Editor.user.js [hunk 1]
```diff
@@ -3167,8 +3168,8 @@
                 'vertical-align': 'bottom',
                 'margin-left': '5px',
                 'font-size': '12px',
-                'color': '#424242',
-                'background': '#eee',
+                'color': 'var(--white)',
+                'background': 'var(--black-800)',
```
Comment (Tyler-H) [inline Magic™Editor.user.js:3172]: The `--black-800` variable is `#e7e8eb` for dark mode, not `#eeeeee`. For light, it's `#242729`. Likewise `--white` is `#2d2d2d` in dark mode and `#ffffff` in light mode, not `#424242`. Are these intended to be one-for-one replacements to bring the values in line with the variable system used in Stacks.css or are they intentionally different as a stylistic choice?

File: Magic™Editor.user.js [hunk 2]
```diff
@@ -78,9 +79,9 @@
             //blockquotes
             //        https://regex101.com/r/fU5lE6/1
             "quote":  /^\>(?:(?!\n\n)[^])+/gm,
-            //inline code
-            //        https://regex101.com/r/9JOfKb/1/
-            "inline": /(?:```(?:[^`](?!\n\n))+?```|`(?:[^`](?!\n\n))+?`)/g,
+            //code surrounded by backticks
+            //        https://regex101.com/r/8tZD3i/2
+            "backickCode": /(?:(?:^(`{3,})[^]+?\1)|(`+)(?:\\`|[^`](?!\n\n))+\2)/gm,
```
Comment (Tyler-H) [inline Magic™Editor.user.js:84]: Typo: should be `backtickCode`.

File: Magic™Editor.user.js [hunk 3]
```diff
@@ -57,7 +58,7 @@
             //    badphrases which relies on "_xPlacexHolderx" starting a placeholder.
             "auto":        "_xPlacexHolderxAutoxInsertxTextxPlacexHolderx_",
             "quote":       "_xPlacexHolderxBlockxQuotexPlacexHolderx_",
-            "inline":      "_xPlacexHolderxCodexInlinexPlacexHolderx_",
+            "backickCode": "_xPlacexHolderxCodexPreserveBlockxPlacexHolderx_",
```
Comment (Tyler-H) [inline Magic™Editor.user.js:61]: Typo: should be `backtickCode`.

File: Magic™Editor.user.js [hunk 4]
```diff
@@ -3134,7 +3135,7 @@
...
+                '            <path style="fill:var(--black-600)" d="m 263,680 c -1,-1 -12,-11 -23,-22 l -21,-21 -1,-3 -1,-3 2,-4 2,-4 128,-128 128,-128 5,0 5,0 23,23 23,23 0,6 0,6 -128,128 -128,128 -4,1 -4,1 -3,-1 z m 237,-263 c 0,-1 -15,-17 -18,-18 l -2,-1 -28,28 c -15,15 -28,28 -28,28 0,0 4,5 10,10 l 10,10 28,-28 c 15,-15 28,-28 28,-29 z m 9,110 c -1,-4 -2,-8 -3,-9 l -1,-2 -8,-2 c -10,-3 -10,-3 1,-7 l 9,-3 2,-8 2,-8 1,0 1,0 2,8 2,8 9,3 9,3 -1,1 c -1,1 -4,2 -8,3 l -7,2 -3,8 -3,8 -1,0 -1,0 -2,-7 z M 323,470 c -0,-1 -3,-8 -5,-16 -3,-8 -5,-16 -6,-17 l -1,-2 -4,-1 c -2,-1 -10,-3 -18,-6 l -13,-4 2,-1 c 1,-1 9,-4 18,-6 l 16,-5 2,-4 c 1,-2 3,-10 6,-18 l 4,-14 1,10e-4 1,10e-4 5,17 c 3,9 5,17 6,18 l 1,1 16,5 c 14,4 20,7 19,8 -0,0 -8,3 -17,5 l -17,5 -1,2 c -1,2 -11,30 -11,33 0,2 -2,2 -3,1 z m -63,-63 c -0,-1 -2,-5 -3,-8 l -2,-7 -7,-2 c -4,-1 -8,-3 -8,-3 l -1,-1 9,-3 9,-3 2,-8 2,-8 1,-0 1,-0 2,8 2,8 9,3 9,3 -1,1 c -1,1 -5,2 -9,4 l -7,2 -3,8 -3,8 -1,0 c -1,0 -1,-1 -2,-2 z m 123,-5 c -1,-4 -2,-7 -3,-8 l -1,-2 -6,-2 c -3,-1 -7,-2 -8,-3 l -2,-2 9,-3 9,-3 2,-8 2,-8 1,0 1,0 2,8 2,8 9,3 9,3 -1,1 c -1,1 -4,2 -8,3 l -7,2 -3,8 -3,8 -1,0 -1,0 -2,-7 z"/>' +
```
Comment (Tyler-H) [inline Magic™Editor.user.js:3138]: `-black-600` is `#c4c8cc` in dark mode and `#535a60` in light mode. Since this is replacing a value of `#444444`, I'm wondering if you intended it to be an exact replacement or intentionally different for stylistic reasons.

File: Magic™Editor.user.js [hunk 5]
```diff
@@ -3304,10 +3305,10 @@
 
         App.pipeMods.edit = function(data) {
             App.funcs.popOriginals();
-
+            var defaultBgColor = App.selections.body.css("background-color");
             // Visually confirm edit - SE makes it easy because the jQuery color animation plugin seems to be there by default
-            App.selections.body.animate({ backgroundColor: '#c8ffa7' }, 10);
-            App.selections.body.animate({ backgroundColor: '#fff' }, 1000);
+            App.selections.body.animate({ backgroundColor: retrieveCSSVariable("--green-200") }, 10);
```
Comment (Tyler-H) [inline Magic™Editor.user.js:3310]: Likewise here `--green-200` in Stacks.css is `#487455`, not `#c8ffa7`. Same question as above; these are *very* different colors from one another.

File: Magic™Editor.user.js [hunk 6]
```diff
@@ -3494,21 +3495,20 @@
         // This is the styling for the diff output.
         $('body').append('<style>' +
                          '.difftitle {' +
-                         '    color: rgb(34, 34, 34);' +
+                         '    color: var(--black);' +
```
Comment (Tyler-H) [inline Magic™Editor.user.js:3498]: Again, here `rgb(34, 34, 34)` is the same as `#222222`; the Stacks.css `--black` variable is `#ffffff` in dark mode and `#0c0d0e` in light.

File: Magic™Editor.user.js [hunk 7]
```diff
@@ -3494,21 +3495,20 @@
...
                          '    font-size: 24px;' +
                          '    font-weight: normal;' +
                          '    line-height: 36px;' +
                          '    margin-bottom: 12px;' +
                          '}' +
                          '.diffbody {' +
                          '    white-space: pre-wrap;' +
-                         '    font-family: "courier new", "lucida sans typewriter", mono, monospace' +
```
Comment (Tyler-H) [inline Magic™Editor.user.js:3505]: Why was this removed rather than replaced? I would expect this to be `Consolas, Menlo, Monaco, Lucida Console, Liberation Mono, DejaVu Sans Mono, Bitstream Vera Sans Mono, Courier New, monospace, sans-serif` to match the site's diff font-family.

File: Magic™Editor.user.js [hunk 8]
```diff
@@ -3494,21 +3495,20 @@
...
                          '}' +
                          '.add {' +
-                         '    background: #CFC;' +
+                         '    background: var(--green-100);' +
```
Comment (Tyler-H) [inline Magic™Editor.user.js:3508]: `--green-100` is equal to `#3d6048` in dark mode or `#cae8d4` in light, not `#CCFFCC`.

File: Magic™Editor.user.js [hunk 9]
```diff
@@ -3494,21 +3495,20 @@
...
                          '}' +
                          '.del {' +
-                         '    background: #FCC;' +
+                         '    background: var(--red-700);' +
```
Comment (Tyler-H) [inline Magic™Editor.user.js:3511]: `--red-700` is `#ac2726` for light or `#de7176` for dark mode, not `#FFCCCC`

File: Magic™Editor.user.js (hunk 2 above)
Comment (Tyler-H) [reply Magic™Editor.user.js:84]: The regex replacement appears to be correct

**Review by Tyler-H — COMMENTED** [review]
Did not do a deep review of the JS aspects of this or test it, just reviewed what I could see from the code here. Also because this is my first review on this and I'm not officially a maintainer of the script, I'm posting this as a comment rather than as 'request changes', but I do request you respond to the comments and fix the "backick" typos, at the very least.

File: Magic™Editor.user.js (hunk 5 above)
Comment (PurpleMagick) [reply Magic™Editor.user.js]: This one strikes a balance between light and dark mode because there is no equivalent that works in both. I went with a green that is different for light mode but still usable and also works in dark mode. The colour is used as a brief highlight of the editor background to signify it has been done. It is there for just a second, so I do not think the exact shade of green is too important to keep. The point is a short signal for completion.

File: Magic™Editor.user.js (hunk 6 above)
Comment (PurpleMagick) [reply Magic™Editor.user.js:3500]: The actual title colour is set to `--black-700` - I will use that instead.

File: Magic™Editor.user.js (hunk 7 above)
Comment (PurpleMagick) [reply Magic™Editor.user.js:3505]: It's removed to inherit the font-family from the site's CSS, rather than override it. Will change it to `var(--ff-mono)` to match the diff style.

File: Magic™Editor.user.js (hunk 4 above)
Comment (PurpleMagick) [reply Magic™Editor.user.js:3138]: More of stylistic choice. The colour determines the icon for the magic wand button. `--black-800` looks contrasting enough for both light and dark theme. Moreover, changing to different `--black-N` options around this has very little visible difference:

Light theme:
-  [`--black-500`](https://i.imgur.com/iR4kb6C.png) 
-  [`--black-600` (current)](https://i.imgur.com/ve4hB4Y.png) 
-  [`--black-700`](https://i.imgur.com/kUVJwMw.png) 

Dark theme:

- [`--black-500`](https://i.imgur.com/5twYMXR.png)
- [`--black-600` (current)](https://i.imgur.com/8yWINC9.png)
- [`--black-700`](https://i.imgur.com/6n88eKi.png)

With `--black-600` it looks good enough as a button. I am not sure choosing a different predefined colour would make much of a difference but I would be happy to change, if needed.

File: Magic™Editor.user.js (hunk 1 above)
Comment (PurpleMagick) [reply Magic™Editor.user.js:3172]: These colours govern the background and text colour of the number of replacements shown after the run (next to the magic wand button(:

- [Light theme](https://i.imgur.com/4k3bOyr.png)
- [Dark theme](https://i.imgur.com/xHWK6P3.png)

There were no direct equivalents in SE's colour palette. And choosing too close values like `--black-200` + `--black-400` or `--black-600` + `--black-800` makes the choice look OK in one theme but not the other. I opted for a much starker contrast to satisfy both themes.

File: Magic™Editor.user.js (hunk 8 above)
Comment (PurpleMagick) [reply Magic™Editor.user.js:3511]: `--green-100` is consistent with the background colour used for diffs in revisions. Matched in the CSS rule for `span.diff-add`

File: Magic™Editor.user.js (hunk 9 above)
Comment (PurpleMagick) [reply Magic™Editor.user.js:3514]: `--red-700` is consistent with the background colour used for diffs in revisions. Matched in the CSS rule `span.diff-delete`

**Review by makyen — COMMENTED** [review]
The only real change that I felt should be made after all the fixes Tyler-H identified and were fixed was that the version should have been 1.8.0, but that's mostly confusion due to not using SemVer and x.x.x format. It's also something that can easily be done by us and shouldn't prevent merging.

**Comment by makyen** [comment]
Thanks for doing this. This is **way** too long for merging to be delayed. I'm sorry it ended up taking so long.
